src/Components/Web.Extensions/src/ProtectedBrowserStorage/ProtectedBrowserStorage.cs

```diff
@@ -0,0 +1,158 @@
+        /// <returns>A <see cref="ValueTask"/> representing the completion of the operation.</returns>
+        public ValueTask<ProtectedBrowserStorageResult<TValue>> GetAsync<TValue>(string key)
+            => GetAsync<TValue>(CreatePurposeFromKey(key), key);
+
+        /// <summary>
+        /// <para>
+        /// Asynchronously retrieves the specified data.
+        /// </para>
+        /// </summary>
+        /// <param name="purpose">A string that defines a scope for the data protection. The protected data can only be unprotected if the same purpose was previously specified when calling <see cref="SetAsync(string, string, object)"/>.</param>
+        /// <param name="key">A <see cref="string"/> value specifying the name of the storage slot to use.</param>
+        /// <returns>A <see cref="ValueTask"/> representing the completion of the operation.</returns>
+        public async ValueTask<ProtectedBrowserStorageResult<TValue>> GetAsync<TValue>(string purpose, string key)
+        {
+            var protectedJson = await GetProtectedJsonAsync(key);
+
+            return protectedJson == null ?
+                new ProtectedBrowserStorageResult<TValue>(false, default) :
+                new ProtectedBrowserStorageResult<TValue>(true, Unprotect<TValue>(purpose, protectedJson));
+        }
+
+        /// <summary>
+        /// Asynchronously deletes any data stored for the specified key.
+        /// </summary>
+        /// <param name="key">A <see cref="string"/> value specifying the name of the storage slot whose value should be deleted.</param>
+        /// <returns>A <see cref="ValueTask"/> representing the completion of the operation.</returns>
+        public ValueTask DeleteAsync(string key)
+            => _jsRuntime.InvokeVoidAsync($"{JsFunctionsPrefix}.delete", _storeName, key);
+
+        private string Protect(string purpose, object value)
+        {
+            var json = JsonSerializer.Serialize(value, options: JsonSerializerOptionsProvider.Options);
+            var protector = GetOrCreateCachedProtector(purpose);
+
+            return protector.Protect(json);
+        }
+
+        private TValue Unprotect<TValue>(string purpose, string protectedJson)
+        {
+            var protector = GetOrCreateCachedProtector(purpose);
+            var json = protector.Unprotect(protectedJson);
+
+            return JsonSerializer.Deserialize<TValue>(json, options: JsonSerializerOptionsProvider.Options)!;
+        }
+
+        private ValueTask SetProtectedJsonAsync(string key, string protectedJson)
+           => _jsRuntime.InvokeVoidAsync($"{JsFunctionsPrefix}.set", _storeName, key, protectedJson);
+
+        private ValueTask<string> GetProtectedJsonAsync(string key)
+            => _jsRuntime.InvokeAsync<string>($"{JsFunctionsPrefix}.get", _storeName, key);
+
+        // IDataProtect isn't disposable, so we're fine holding these indefinitely.
```

Comment:
> perhaps we get away caching the default protector

Is there a default protector? If no purpose is specified, then a purpose is derived from the storename and key.

If the protectors are very cheap to create then yes it would be better not to cache them at all. Or finding some way to cache them inside the circuit so they can't live longer than the user's connection (though this would require some new DI service for per-circuit state).

Reply:
Creating a new protector isn't terribly expensive, but it's not free. If there's a protector that you know you're going to use over and over again, may as well cache it. But if it's a protector that you might end up using only a small handful of times, meh.

@pranavkm gave a good rule of thumb. Cache the _default_ protector, and create the rest on-the-fly. This is the pattern that ASP.NET full framework uses in its data protection mechanism, which was the foundation for ASP.NET Core's data protection mechanism. Basically, the "root" protector (where the purpose is the type name) is cached somewhere, and any child protectors derived from it are created on-demand.

Take a look at all the readonly statics at the beginning of https://referencesource.microsoft.com/System.Web/R/6395d736fd374b03.html for more info. In your case you probably can't use statics since you have a DI system, but you should still be able to cache it within your singleton's instance.

Reply:
These are registered as scoped, so you are essentially going to get one per circuit. I suggest you create an instance on the constructor and stuff it on an instance field. That should be enough. @GrabYourPitchforks can confirm, but there's no issue grabbing to the IDataProtector for a long time, auth does that.

As to how to make it unique per user, I can't think of a good way to achieve it and it can cause problems. The main drawback would be that someone could hijack your protected storage (if you get XSSd) copy the values into their own app session and somehow trigger the app to read them?

I'm not sure how much we care about that given that you have already been XSSed. @GrabYourPitchforks do you have any thoughts?

The problem that I see otherwise is that if you login/logout you can run into issues about not being able to read an existing payload, and I don't know if that's desirable or not.